keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1428 @@
+  This kubelet will be unaware of the volume, and beginning running the pod
+  without mounting any certificates.  If the application inside the pod is
+  expecting the certificates to be present, it will probably crash.
+* (`kube-apiserver>=1.32`, `kubelet<1.32`, `kube-controller-manager>=1.32`):
+  Same as the above case.
+* (`kube-apiserver>=1.32`, `kubelet>=1.32`, `kube-controller-manager<1.32`):
+  kube-apiserver will admit pod specs with PodCertificate projected volumes.
+  When scheduled, kubelet will successfully create a PodCertificateRequest
+  object for the pod.  However, these PodCertificateRequests will hang in an
+  unissued state forever, which means the pods will remain stuck at volume set
+  up. `kubernetes.io/kube-apiserver-client-pod`, kube-controller-manager will
+  not act on it.  Workloads that were trying to create these pods will be stuck
+  at pod setup.
+* (`kube-apiserver>=1.32`, `kubelet>=1.32`, `kube-controller-manager>=1.32`):
+  The feature works normally.
+
+Once the feature is launched, we need to consider feature evolution.
+
+One possible example is support for a new key type.  Supporting a new key type
+requires updates in all three components:
+* Kubelet will need support for generating keys of the new type.
+* Kube-apiserver will need support for deserializing the new key type and
+  validating the proof-of-possession.
+* Kube-controller-manager (and third-party signing controllers) will need
+  support for issuing certificates to the new key type.
+
+If support for the new key type is introduced in version N, then we have the
+following skew cases:
+* (`kube-apiserver>=N`, `kubelet<N`, `kube-controller-manager<N`):
+  kube-apiserver will admit pods with PodCertificate projected volumes that
+  specify the new key type, but these pods will fail to start when actually
+  scheduled onto a node.  Kubelet will continually return an error during volume
+  set up, when it validates the specified key type against its list of supported
+  key types.
+* (`kube-apiserver>=N`, `kubelet<N`, `kube-controller-manager>=N`): Same as the
+  previous case.
+* (`kube-apiserver>=N`, `kubelet>=N`, `kube-controller-manager<N`):
+  kube-apiserver will admit pods with PodCertificate projected volumes that
+  specify the new key type.  These pods will still fail to start when scheduled
+  onto a node.  Kubelet will successfully create PodCertificateRequest objects,
+  but kube-controller-manager will continually reject them because it does not
+  understand the new key type.  Kubelet will reflect this by returning a
+  permanent error from volume set up.
+* (`kube-apiserver>=N`, `kubelet>=N`, `kube-controller-manager>=N`): No
+  problems, everything works.
+
+Based on the above analysis, we may want to consider making kube-scheduler aware
+of which kubelet versions support PodCertificate projected volumes (and which
+key types are supported by version), and try to steer pods towards nodes that
+can successfully mount them.
```

Comment:
We don't do that kind of checks anywhere I'm aware of, so I'd advise against it.

Reply:
Rather make sure to provide sufficient information in the logs or through events which would allow a user to identify such a situation.

Reply:
Removed that paragraph.